source/routing/journey_pattern_container.cpp

```diff
@@ -0,0 +1,160 @@
+/* Copyright © 2001-2015, Canal TP and/or its affiliates. All rights reserved.
+  
+This file is part of Navitia,
+    the software to build cool stuff with public transport.
+ 
+Hope you'll enjoy and contribute to this project,
+    powered by Canal TP (www.canaltp.fr).
+Help us simplify mobility and open public transport:
+    a non ending quest to the responsive locomotion way of traveling!
+  
+LICENCE: This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+   
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Affero General Public License for more details.
+   
+You should have received a copy of the GNU Affero General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+  
+Stay tuned using
+twitter @navitia 
+IRC #navitia on freenode
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+
+#include "journey_pattern_container.h"
+#include "type/pt_data.h"
+
+namespace navitia { namespace routing {
+
+namespace nt = navitia::type;
+
+template<> std::vector<const nt::DiscreteVehicleJourney*>&
+JourneyPattern::get_vjs<nt::DiscreteVehicleJourney>() {
+    return discrete_vjs;
+}
+template<> std::vector<const nt::FrequencyVehicleJourney*>&
+JourneyPattern::get_vjs<nt::FrequencyVehicleJourney>() {
+    return freq_vjs;
+}
+
+void JourneyPatternContainer::init(const nt::PT_Data& pt_data) {
+    map.clear();
+    jps.clear();
+    jpps.clear();
+    jps_from_route.assign(pt_data.routes);
+    for (const auto* jp: pt_data.journey_patterns) {
+        for (const auto& vj: jp->discrete_vehicle_journey_list) { add_vj(*vj); }
+        for (const auto& vj: jp->frequency_vehicle_journey_list) { add_vj(*vj); }
+    }
+}
+
+JourneyPatternContainer::JpKey JourneyPatternContainer::make_key(const nt::VehicleJourney& vj) {
+    JourneyPatternContainer::JpKey key;
+    key.route_idx = RouteIdx(*vj.journey_pattern->route);
+    key.is_freq = dynamic_cast<const nt::FrequencyVehicleJourney*>(&vj) != nullptr;
+    for (const auto& st: vj.stop_time_list) {
+        key.jpp_keys.emplace_back(SpIdx(*st.journey_pattern_point->stop_point),
+                                  st.local_traffic_zone,
+                                  st.properties);
+    }
+    return key;
+}
+
+static bool st_are_equals(const nt::VehicleJourney& vj1, const nt::VehicleJourney& vj2) {
```

Comment:
Will it be better to check  `vj1.stop_time_list.size() == vj2.stop_time_list.size()` at the first place?